records/2024-04-30.magic-link-security-team-proposal.md

```diff
@@ -0,0 +1,112 @@
+# Notify 2FA using Magic Links and a PKCE-like mechanism
+
+Using a mechanism similar to OAuth 2.0's Proof Key for Code Exchange (PKCE) for the verification of magic links can be a robust and secure alternative to the [proposed solution](https://github.com/cds-snc/notification-adr/pull/64/files), even if Notify isn't using OAuth itself.
+This approach leverages the principles of PKCE, originally designed to secure OAuth against interception attacks, and can similarly enhance the security of any process where tokens or links are exchanged.
+Some downsides of the already proposed solution (with Argon2D), that the PKCE-like mechanism could solve:
+* Policy and Legal implications of collecting users' browser attributes and / or IP addresses
+* An increase in AWS costs, and system performance, related to running the Argon2d hashing algorithm for each login
+* Level of effort - this solution would likely not be trivial to implement, and/or maintain in the future.
+
+# Concept Overview
+
+PKCE uses a pair of related tokens, the code_verifier and the code_challenge, to secure the exchange:
+* Code Verifier: A high-entropy cryptographic random string generated by the client, which is kept secret and used later to verify the authenticity of the request.
+* Code Challenge: A derived version of the code_verifier, usually created by hashing the code_verifier and sending it over the network (in PKCE, this is sent to the authorization server during the initial request).
+
+# Applying PKCE-like mechanism to Magic Links
```

Comment:
That is a good suggestion in the context of magic links. Our magic link differs a bit though in the sense that the user had to authenticate before and we left a cookie on the browser along with a session ID that both backend and front end should know about (on top of my head -- if I am wrong, tell me). 

Hence I have the feeling that this process achieves the same, i.e. to identify and bound the user' session between front-end and back-end, in the context of password-less magic links where there were no prior contact. Maybe the current implementation we have might be improved with the suggested code challenge, I have the impression though our workflow could be different given we don't have "true" password-less magic link.

Does that make sense? Are some of my assumptions wrong? And if we implement this logic for magic links, should we adapt it to our password present workflow and how?

Reply:
I tend to think you are right here, @jimleroyer.  We achieve basically the same thing with the cookie that is provided once the user provides their username and password.  

Given that is the case, would this be necessary?